### PR TITLE
Several bug fixes, metrics api, and tests

### PR DESCRIPTION
**Pull request by geek**
* Fix bug where 503 retries didn't work
* Fix deadlock issue in login retries
* Enhance error logging to include more details
* Add several unit tests
* Adds metrics API support

### CODE REVIEW

**Comment by virgilm** [comment]
Getting the following error when running tests:

> ==> Building the Datera Golang SDK
go: downloading gopkg.in/h2non/gock.v1 v1.0.15
go: downloading github.com/tjcelaya/go-datera v1.2.1
go: downloading github.com/google/go-cmp v0.4.0
go: downloading github.com/h2non/parth v0.0.0-20190131123155-b4df798d6542
go: downloading golang.org/x/xerrors v0.0.0-20191204190536-9bdfabe68543
github.com/Datera/go-sdk/pkg/dsdk
pkg/dsdk/remote_provider.go:254:3: struct field tag `json:"snapshot_app_instance_path" "mapstructure:"snapshot_app_instance_path"` not compatible with reflect.StructTag.Get: bad syntax for struct tag key
github.com/Datera/go-sdk/tests_test
vet: tests/sdk_test.go:304:15: ErrRetryTimeout not declared by package dsdk
make: *** [compile] Error 2
Build step 'Execute shell' marked build as failure
Finished: FAILURE
`

See http://eco-jenkins.daterainc.com:8080/job/K8S-go-sdk-validation/33/console

File: pkg/dsdk/remote_provider.go [hunk 1]
```diff
@@ -201,3 +233,42 @@ func (e *RemoteProvider) Reload(ro *RemoteProviderReloadRequest) (*RemoteProvide
...
+	Uuid               string `json:"uuid" mapstructure:"uuid"`
+	RemoteProviderUuid string `json:"remote_provider_uuid" mapstructure:"remote_provider_uuid"`
+	AppInstanceUuid    string `json:"app_instance_uuid" mapstructure:"app_instance_uuid"`
+	OpState            string `json:"op_state" mapstructure:"op_state"`
+	OpType             string `json:"op_type" mapstructure:"op_type"`
+	PercentDone        int    `json:"percent_done" mapstructure:"percent_done"`
+	TotalTasksDone     int    `json:"total_tasks_done" mapstructure:"total_tasks_done"`
+	TotalTasksIssued   int    `json:"total_tasks_issued" mapstructure:"total_tasks_issued"`
+	References         struct {
+		SnapshotAppInstancePath string `json:"snapshot_app_instance_path" "mapstructure:"snapshot_app_instance_path"`
```
[the remark this answers is not in the document]
Comment (virgilm) [reply pkg/dsdk/remote_provider.go]: I fixed the UT here: https://github.com/Datera/go-sdk/tree/test-geek-pr. Please check it out.

**Comment by virgilm** [comment]
I fixed the UT here: https://github.com/Datera/go-sdk/tree/test-geek-pr. Please check it out.

**Comment by virgilm** [comment]
Please see other changes required at https://github.com/Datera/go-sdk/tree/test-geek-pr. Happy to merge this PR once those are addressed.

**Comment by geek** [comment]
@virgilm updated with those changes

**Comment by virgilm** [comment]
@geek I tried to merge this, but I am running into verification errors. Looking at it, but I might need to revert this change. See full log.
[consoleText.txt](https://github.com/Datera/go-sdk/files/4624631/consoleText.txt)

**Pull request merged**